### PR TITLE
DRAFT: Parallel Experiment Designer

### PR DESCRIPTION
**Pull request by byte-sculptor**
Looking through Greg's changes

### CODE REVIEW

File: source/Mlos.Python/mlos/OptimizerEvaluationTools/ObjectiveFunctionBase.py [hunk 1]
```diff
@@ -56,6 +57,11 @@ def evaluate_point(self, point: Point) -> Point:
         # If evaluate_point is not implemented in the subclass, we can make it work like so:
         #
         point_df = point.to_dataframe()
+
+        # TODO remove non numeric numbers
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/OptimizerEvaluationTools/ObjectiveFunctionBase.py:61]: undo

File: source/Mlos.Python/mlos/Optimizers/BayesianOptimizer.py [hunk 2]
```diff
@@ -76,14 +77,27 @@ def __init__(
...
-        assert self.optimizer_config.experiment_designer_implementation == ExperimentDesigner.__name__
-        self.experiment_designer = ExperimentDesigner(
-            designer_config=self.optimizer_config.experiment_designer_config,
-            optimization_problem=self.optimization_problem,
-            pareto_frontier=self.pareto_frontier,
-            surrogate_model=self.surrogate_model,
-            logger=self.logger
-        )
+
+        # #TODO, great example of Python reflection
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/BayesianOptimizer.py:82]: ha ha

File: source/Mlos.Python/mlos/Optimizers/BayesianOptimizer.py [hunk 3]
```diff
@@ -130,11 +144,11 @@ def compute_surrogate_model_goodness_of_fit(self):
     def get_optimizer_convergence_state(self):
         return self._optimizer_convergence_state
 
-    def get_all_observations(self):
+    def get_all_observations(self) -> [pd.DataFrame, pd.DataFrame, pd.DataFrame]:
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/BayesianOptimizer.py]: Make this a Tuple[...]

File: source/Mlos.Python/mlos/Optimizers/BayesianOptimizer.py [hunk 4]
```diff
@@ -218,3 +257,31 @@ def predict(self, parameter_values_pandas_frame, t=None, context_values_pandas_f
             objective_name = self.optimization_problem.objective_names[0]
 
         return self.surrogate_model.predict(feature_values_pandas_frame)[objective_name]
+
+    @trace()
+    def _update_pareto(self):
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/BayesianOptimizer.py:263]: This should be configurable.

File: source/Mlos.Python/mlos/Optimizers/ExperimentDesigner/ExperimentDesigner.py [hunk 5]
```diff
@@ -3,6 +3,8 @@
 # Licensed under the MIT License.
 #
 import numpy as np
+import pandas as pd
+import pandas as pd
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/ExperimentDesigner/ExperimentDesigner.py:7]: remove

File: source/Mlos.Python/mlos/Optimizers/ExperimentDesigner/ExperimentDesigner.py [hunk 6]
```diff
@@ -191,3 +193,12 @@ def suggest(self, context_values_dataframe=None, random=False):
         except UnableToProduceGuidedSuggestionException:
             self.logger.info("Failed to produce guided suggestion. Producing random suggestion instead.")
             return self.optimization_problem.parameter_space.random()
+
+    def add_pending_suggestion(self, suggestion: Point):
+        ...
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/ExperimentDesigner/ExperimentDesigner.py:198]: ... -> pass

Also, create a suitable ExperimentDesignerInterface

**Comment by amueller** [comment]
I'd love to see this move forward. Do you want reviews or someone to take over the code or is there some other way I can help?

**Pull request closed without merging**